### ISSUE TITLE
Systems and LazySets

### ISSUE THREAD

**Issue by mforets** [JuliaReach/SX.jl#12]
This issue is a continuacion of #7. The next step is to allow different a value for the system type assumed for each mode, eg. linear affine through `ST = ConstrainedLinearControlContinuousSystem`.

If we rely eg. on SymEngine the flow equations can be transformed from the Julia expressions into the matrix `A`, `B` for `x' = Ax + Bu`. 

Special care is taken for constants, like the `-1` in `x' = v - 1`. Here, we can add an extra row of `B` with the coefficient number, and a corresponding new input in the vector of `U`',s with a constant value of `1`. If `LazySets` is used, this can be a set with one element, that is, a `Singleton`

The invariants also could be transformed to `LazySets` objects, eg. `u1 >= 0` to a half-space and so on.

It is not clear to me how to deal with the reset maps and switchings. Sometimes they take the form of assignments, or equations or inequalities. One approach that has been suggested is to use symbolic functions from SymEngine, and a custom type for equality.

**Comment by schillic** [comment]
> It is not clear to me how to deal with the reset maps and switchings. Sometimes they take the form of assignments, or equations or inequalities.

I would expect that SpaceEx has a required format. The manual mentions the form `var' == new_expr`. For convenience you could still support `var := new_expr`.
I have not seen inequalities here, but the meaning would probably be a nondeterministic assignment.

Altogether, I would go for a mapping from a variable to a set. The set is either a `Hyperplane` (equalities) or a `HalfSpace` (inequalities) (I hope that is correct).

**Comment by mforets** [comment]
> I would expect that SpaceEx has a required format. The manual mentions the form var' == new_expr. For convenience you could still support var := new_expr.

yes. that's been clarified in the last meeting with @nikos-kekatos . basically, in assignments `var' == ...` really defines the discrete system law, like `var' = var(after the switch) = f(var before the switch, ...)` 

> Altogether, I would go for a mapping from a variable to a set. The set is either a Hyperplane (equalities) or a HalfSpace (inequalities) (I hope that is correct).

agreed about that. worth mentioning that if one is only interested in concrete polyhedra (Polyhedra.jl), that's an option that could be added later (like concrete/symblic).

**Comment by mforets** [comment]
> The set is either a Hyperplane (equalities) or a HalfSpace (inequalities) (I hope that is correct).

Actually in general the set will be defined with several subexpressions, in that case i think it is convenient to use an intersection array ( https://github.com/JuliaReach/LazySets.jl/issues/323 ).

**Comment by mforets** [comment]
The bouncing ball has a guard of the form `ax < b`, that is with a strict inequality:

```julia
  "guards"           => Array{Expr,1}[Expr[:(x = 0), :(v < 0)]]
``

I think that we should use `HalfSpace` because we are using floating point. Do you agree? Does SpaceEx treat in any special case the strict inequalities vs non-strict ones?

**Comment by schillic** [comment]
> I think that we should use HalfSpace because we are using floating point. Do you agree? Does SpaceEx treat in any special case the strict inequalities vs non-strict ones?

Agreed. As most LP solvers, GLPK (used in SpaceEx) [seems to only allow for non-strict inequalities](https://lists.gnu.org/archive/html/help-glpk/2004-09/msg00029.html). Hence I assume that SpaceEx has the same treatment, but I am not sure.

**Comment by nikos-kekatos** [comment]
@schillic Nice catch with the strict/non-strict inequalities! So, the STC &LGG scenarios rely on non-strict inequalities. That means that any $<$, $>$ in a model would be transformed to $\leq$, $\geq$ respectively by SpaceEx. The PHAver differentiates between strict and non-strict inequalities as it employs PPL.

**Comment by nikos-kekatos** [comment]
Some information about SX modeling language can be found at  this [link](http://spaceex.imag.fr/sites/default/files/spaceex_modeling_language_0.pdf)

**Comment by nikos-kekatos** [comment]
Note that for resets  x=1, x:=1 and x'==1 are accepted.  However, x==1 is not correct (as == is typically used for checks/comparison but not for assignements).

**Comment by nikos-kekatos** [comment]
For flows, both x'==1 and x:=1 correspond to the first derivative of x wrt time.
